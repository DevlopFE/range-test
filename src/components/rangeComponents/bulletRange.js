import React, {useEffect} from 'react'

export const BulletRange = React.forwardRef( ({bulletRef, sign, position, ...props}) => {
  useEffect(() => {
    if(bulletRef.current){
      bulletRef.current.style.left = `${position}px`;
    }
  }, [bulletRef, position]);

  return (
    <div ref={bulletRef} {...props} />
  )
})
