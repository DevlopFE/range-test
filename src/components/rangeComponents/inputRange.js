import React from 'react'

export const InputRange = React.forwardRef( ({inputRef, position, ...props}) => {
  return (
    <input 
      // type="number"
      ref={inputRef}
      value={position}
      {...props}
      readOnly
    />  
  )
})

