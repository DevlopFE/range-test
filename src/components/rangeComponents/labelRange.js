import React from 'react'

export const LabelRange = ({ sign, ...props}) => {
  
  return (
    <div {...props} >{sign}</div>
  )
}
