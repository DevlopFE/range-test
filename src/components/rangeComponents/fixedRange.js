import React, {useState, useEffect, useCallback, useRef} from 'react'
import { InputRange } from './inputRange';
import { BulletRange } from './bulletRange';
import { LabelRange } from './labelRange';

export const FixedRange = ({ values }) => {

  const INIT_POSITION_MIN = values[0];
  const INIT_POSITION_MAX = values[values.length - 1];
  const LIMIT = 300
  const minColor='#282c34';
  const maxColor='rgb(228, 164, 108)';
  const isDraggingMin = useRef(false);
  const isDraggingMax = useRef(false);
  const dragHeadRefMin = useRef();
  const dragHeadRefMax = useRef();
  const inputMin = useRef();
  const inputMax = useRef();
  const [positionMin, setPositionMin] = useState(INIT_POSITION_MIN);
  const [positionMax, setPositionMax] = useState(INIT_POSITION_MAX);
  const steps = values
  const [step, setStep] = useState(0);
  const stepsRight = values
  const [stepRight, setStepsRight] = useState(stepsRight.length);
    

  
  const onMouseDownLeft = useCallback((e) => {
    if (dragHeadRefMin.current && dragHeadRefMin.current.contains(e.target)) {
      isDraggingMin.current = true;
      dragHeadRefMin.current.style.backgroundColor = minColor;
      dragHeadRefMin.current.style.transition = "transform .2s";
      dragHeadRefMin.current.style.transform = "scale(1.2)";
    }
  }, []);

  const onMouseDownRight = useCallback((e) => {
    if ( dragHeadRefMax.current && dragHeadRefMax.current.contains(e.target)){
      isDraggingMax.current = true;
      dragHeadRefMax.current.style.backgroundColor = maxColor;
      dragHeadRefMax.current.style.transition = "transform .2s";
      dragHeadRefMax.current.style.transform = "scale(1.2)";
    }
  }, []);


  const onMouseUpLeft = useCallback((e) => {
    if (isDraggingMin.current) {
      isDraggingMin.current = false;
      dragHeadRefMin.current.style.backgroundColor = minColor;
      dragHeadRefMin.current.style.transition = "transform .2s";
      dragHeadRefMin.current.style.transform = "scale(1)";
    }
  }, []);

  const onMouseUpRight = useCallback(() => {
    if (isDraggingMax.current) {
      isDraggingMax.current = false;
      dragHeadRefMax.current.style.backgroundColor = maxColor;
      dragHeadRefMax.current.style.transition = "transform .2s";
      dragHeadRefMax.current.style.transform = "scale(1)";
    }
  }, []);

  const onMouseMoveLeft = useCallback( (e) => {
    if (isDraggingMin.current) {
      const valueToCheck =e.movementX*5;

          if(valueToCheck > 0 && positionMin < positionMax && step < steps.length-1){
            const newLeftPosition = steps[step+1]
            if(newLeftPosition < positionMax){
            setStep(step => step+1)
            setPositionMin(newLeftPosition);
          }
          }
        if(valueToCheck < 0 && positionMin > INIT_POSITION_MIN && step > 0){
          setStep(step => step-1)
          const newLeftPosition = steps[step]
          setPositionMin(newLeftPosition);
          }

          if(valueToCheck < 0 && step === 0){
            setStep(0)
            const newLeftPosition = steps[step]
            setPositionMin(newLeftPosition);
            }
      }  
    }
  , [positionMin, positionMax, step, steps]);


  const onMouseMoveRight = useCallback( (e) => {
    if (isDraggingMax.current) {
      const valueToCheck =e.movementX*5;

          if(valueToCheck < 0 && positionMax > positionMin){
            const newRightPosition = stepsRight[stepRight-1]
            if(newRightPosition > positionMin){
              setStepsRight(stepRight => stepRight-1)
              setPositionMax(newRightPosition);
            }
          }
        if(valueToCheck > 0 && positionMax > positionMin && stepRight < stepsRight.length){
          setStepsRight(stepRight => stepRight+1)
          const newRightPosition = stepsRight[stepRight]
          setPositionMax(newRightPosition);
          }
      }  
    }
  , [positionMin, positionMax, stepRight, stepsRight]);



  useEffect(() => {
    document.addEventListener("mouseup", onMouseUpLeft);
    document.addEventListener("mousedown", onMouseDownLeft);
    document.addEventListener("mousemove", onMouseMoveLeft);
    return () => {
      document.removeEventListener("mouseup", onMouseUpLeft);
      document.removeEventListener("mousedown", onMouseDownLeft);
      document.removeEventListener("mousemove", onMouseMoveLeft);
    };
  }, [onMouseMoveLeft, onMouseDownLeft, onMouseUpLeft]);

  useEffect(() => {
    document.addEventListener("mouseup", onMouseUpRight);
    document.addEventListener("mousedown", onMouseDownRight);
    document.addEventListener("mousemove", onMouseMoveRight);
    return () => {
      document.removeEventListener("mouseup", onMouseUpRight);
      document.removeEventListener("mousedown", onMouseDownRight);
      document.removeEventListener("mousemove", onMouseMoveRight);
    };
  }, [onMouseMoveRight, onMouseDownRight, onMouseUpRight]);


  

  useEffect(() => {
    if(isDraggingMin.current && positionMin <= INIT_POSITION_MIN ){
      setPositionMin(INIT_POSITION_MIN)
    }
    if(isDraggingMin.current && positionMin >= positionMax  ){
      setPositionMin(positionMax)
      isDraggingMax.current = false;
    }
    if(isDraggingMax.current && positionMax <= positionMin ){
      setPositionMax(positionMin)
      isDraggingMin.current = false;
    }
    if(isDraggingMax.current && positionMax >= INIT_POSITION_MAX ){
      setPositionMax(INIT_POSITION_MAX)
    }
    if(isDraggingMax.current && positionMax <= INIT_POSITION_MIN ){
      setPositionMax(INIT_POSITION_MIN)
    }
  }, [positionMin, positionMax, INIT_POSITION_MIN, INIT_POSITION_MAX])

  return (
    <>
      <div className='range-container'>
        <div className='range-slider'>
          
          <InputRange 
            className='minvalue-input'
            inputRef={inputMin} 
            position={positionMin}
          /> 
          <LabelRange className='minvalue-label' sign={'$'} />
          <BulletRange bulletRef={dragHeadRefMin} minValue={INIT_POSITION_MIN} maxValue={INIT_POSITION_MAX} limit={LIMIT} position={positionMin} className='minvalue-bullet' />


          <InputRange 
            className='maxvalue-input'
            inputRef={inputMax} 
            position={positionMax}
          /> 
          <LabelRange className='maxvalue-label' sign={'$'} />
          <BulletRange bulletRef={dragHeadRefMax} minValue={INIT_POSITION_MIN} maxValue={INIT_POSITION_MAX} limit={LIMIT} position={positionMax} className='maxvalue-bullet' />

        </div>
      </div>
    </>
  )
}
