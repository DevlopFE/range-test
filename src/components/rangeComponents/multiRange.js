import React, {useState, useEffect, useCallback, useRef} from 'react'
import { InputRange } from './inputRange';
import { BulletRange } from './bulletRange';
import { LabelRange } from './labelRange';

export const MultiRange = ({minValue, maxValue}) => {

  const INIT_POSITION_MIN = minValue || 0;
  const INIT_POSITION_MAX = maxValue || 300;
  const LIMIT = 300
  const minColor='#282c34';
  const maxColor='rgb(228, 164, 108)';
  const isDraggingMin = useRef(false);
  const isDraggingMax = useRef(false);
  const dragHeadRefMin = useRef();
  const dragHeadRefMax = useRef();
  const inputMin = useRef();
  const inputMax = useRef();
  const [positionMin, setPositionMin] = useState(INIT_POSITION_MIN);
  const [positionMax, setPositionMax] = useState(INIT_POSITION_MAX);

  
  const onMouseDownLeft = useCallback((e) => {
    if (dragHeadRefMin.current && dragHeadRefMin.current.contains(e.target)) {
      isDraggingMin.current = true;
      dragHeadRefMin.current.style.backgroundColor = minColor;
      dragHeadRefMin.current.style.transition = "transform .2s";
      dragHeadRefMin.current.style.transform = "scale(1.2)";
    }
  }, []);

  const onMouseUpLeft = useCallback(() => {
    if (isDraggingMin.current) {
      isDraggingMin.current = false;
      dragHeadRefMin.current.style.backgroundColor = minColor;
      dragHeadRefMin.current.style.transition = "transform .2s";
      dragHeadRefMin.current.style.transform = "scale(1)";
    }
  }, []);

  const onMouseMoveLeft = useCallback( (e) => {
    if (isDraggingMin.current) {
      setPositionMin((positionMin) => positionMin + e.movementX);
    }
  }, []);

  const onMouseDownRight = useCallback((e) => {
    if (
      dragHeadRefMax.current &&
      dragHeadRefMax.current.contains(e.target)
    ) {
      dragHeadRefMax.current.style.backgroundColor = maxColor;
      dragHeadRefMax.current.style.transition = "transform .2s";
      dragHeadRefMax.current.style.transform = "scale(1.2)";
      isDraggingMax.current = true;
    }
  }, []);

  const onMouseUpRight = useCallback(() => {
    if (isDraggingMax.current) {
      isDraggingMax.current = false;
      dragHeadRefMax.current.style.backgroundColor = maxColor;
      dragHeadRefMax.current.style.transition = "transform .2s";
      dragHeadRefMax.current.style.transform = "scale(1)";
    }
  }, []);

  const onMouseMoveRight = useCallback( (e) => {
    if (isDraggingMax.current) {

      setPositionMax((positionMax) => positionMax + e.movementX);
    }
  }, []);

  const onMouseChangeMin =  (e) => {
    setPositionMin(parseInt(e.target.value));
}

  const onMouseChangeMax =  (e) => {
      setPositionMax(parseInt(e.target.value));
   }

   const handleClickMin = () => {
    inputMin.current.focus()
   }

   const handleClickMax = () => {
    inputMax.current.focus()
   }

  const handleKeyPressMin = (e) => {
    if (e.key === 'Enter') {
      if (inputMin.current.value <= positionMax && inputMin.current.value !== ''){
    e.target.blur()
      }else{
         setPositionMin(INIT_POSITION_MIN);
      }
    } 
  }

  useEffect(() => {
    document.addEventListener("mouseup", onMouseUpLeft);
    document.addEventListener("mousedown", onMouseDownLeft);
    document.addEventListener("mousemove", onMouseMoveLeft);
    return () => {
      document.removeEventListener("mouseup", onMouseUpLeft);
      document.removeEventListener("mousedown", onMouseDownLeft);
      document.removeEventListener("mousemove", onMouseMoveLeft);
    };
  }, [onMouseMoveLeft, onMouseDownLeft, onMouseUpLeft]);

  useEffect(() => {
    document.addEventListener("mouseup", onMouseUpRight);
    document.addEventListener("mousedown", onMouseDownRight);
    document.addEventListener("mousemove", onMouseMoveRight);
    return () => {
      document.removeEventListener("mouseup", onMouseUpRight);
      document.removeEventListener("mousedown", onMouseDownRight);
      document.removeEventListener("mousemove", onMouseMoveRight);
    };
  }, [onMouseMoveRight, onMouseDownRight, onMouseUpRight]);


  useEffect(() => {
    if(isNaN(positionMax)){
     return
    }
  }, [positionMax])

  useEffect(() => {
    if(isNaN(positionMin)){
     return
    }
  }, [positionMin])
  

  useEffect(() => {
    if(positionMin < INIT_POSITION_MIN ){
      setPositionMin(INIT_POSITION_MIN)
      return
    }
    if(positionMin >= positionMax ){
      setPositionMin(positionMax )
      return
    }
    if(positionMax > INIT_POSITION_MAX ){
      setPositionMax(INIT_POSITION_MAX)
      return
    }
    if(positionMax <= positionMin ){
      setPositionMax(positionMax)
      return
    }
  }, [positionMin, positionMax, INIT_POSITION_MIN, INIT_POSITION_MAX])

  return (
    <>
      <div className='range-container'>
        <div className='range-slider'>
          
          <InputRange 
            className='minvalue-input'
            inputRef={inputMin} 
            onMouseChange={onMouseChangeMin}
            handleKeyPress={handleKeyPressMin}
            position={positionMin}
          /> 
          <LabelRange className='minvalue-label' onClick={handleClickMin} sign={'$'} />
          <BulletRange bulletRef={dragHeadRefMin}  position={positionMin} className='minvalue-bullet' />


          <InputRange 
            className='maxvalue-input'
            inputRef={inputMax} 
            onMouseChange={onMouseChangeMax}
            //handleKeyPress={()=> {}}
            position={positionMax}
          /> 
          <LabelRange className='maxvalue-label' onClick={handleClickMax} sign={'$'} />
          <BulletRange bulletRef={dragHeadRefMax}  position={positionMax} className='maxvalue-bullet' />

        </div>
      </div>
    </>
  )
}
