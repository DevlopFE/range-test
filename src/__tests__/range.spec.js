import { render } from '@testing-library/react';
import {Range} from '../components/rangeComponents/range' 

describe('Range', () => {
  it('should be defined', () => {
    expect(Range).toBeDefined();
  });

  it('should render correctly in normal mode', () => {
    const { container } = render(<Range minValue={0} maxValue={100} />);
    expect(container).toMatchSnapshot();
  })

  it('should render correctly in fixed mode', () => {
    const { container } = render(<Range rangeValues={[0,50,100]} />);
    expect(container).toMatchSnapshot();
  })
}
);
