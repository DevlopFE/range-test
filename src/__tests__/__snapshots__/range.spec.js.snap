// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range should render correctly in fixed mode 1`] = `
<div>
  <div
    class="range-container"
  >
    <div
      class="range-slider"
    >
      <input
        class="minvalue-input"
        readonly=""
        value="0"
      />
      <div
        class="minvalue-label"
      >
        $
      </div>
      <div
        class="minvalue-bullet"
        style="left: 0px;"
      />
      <input
        class="maxvalue-input"
        readonly=""
        value="100"
      />
      <div
        class="maxvalue-label"
      >
        $
      </div>
      <div
        class="maxvalue-bullet"
        style="left: 200px;"
      />
    </div>
  </div>
</div>
`;

exports[`Range should render correctly in normal mode 1`] = `
<div>
  <div
    class="range-container"
  >
    <div
      class="range-slider"
    >
      <input
        class="minvalue-input"
        readonly=""
        value="0"
      />
      <div
        class="minvalue-label"
      >
        $
      </div>
      <div
        class="minvalue-bullet"
        style="left: 0px;"
      />
      <input
        class="maxvalue-input"
        readonly=""
        value="100"
      />
      <div
        class="maxvalue-label"
      >
        $
      </div>
      <div
        class="maxvalue-bullet"
        style="left: 200px;"
      />
    </div>
  </div>
</div>
`;
